webhook_redelivery.py

```diff
@@ -0,0 +1,412 @@
+    github_auth: GithubAuthDetails, webhook_address: WebhookAddress, since_seconds: int
+) -> int:
+    """Redeliver failed webhook deliveries since a given time.
+
+    Args:
+        github_auth: The GitHub authentication details used to interact with the Github API.
+        webhook_address: The data to identify the webhook.
+        since_seconds: The amount of seconds to look back for failed deliveries.
+
+    Returns:
+        The number of failed webhook deliveries redelivered.
+
+    """
+    github = _get_github_client(github_auth)
+
+    deliveries = _iter_delivery_attempts(github_client=github, webhook_address=webhook_address)
+    since_datetime = datetime.now(tz=timezone.utc) - timedelta(seconds=since_seconds)
+    failed_deliveries = _filter_for_failed_attempts(
+        deliveries=deliveries, since_datetime=since_datetime
+    )
+    redelivered_count = _redeliver_attempts(
+        deliveries=failed_deliveries, github_client=github, webhook_address=webhook_address
+    )
+
+    return redelivered_count
+
+
+# Github App authentication is not tested in unit tests, but in integration tests.
+def _get_github_client(github_auth: GithubAuthDetails) -> Github:  # pragma: no cover
+    """Get a Github client.
+
+    Args:
+        github_auth: The Github authentication details.
+
+    Returns:
+        The Github client.
+    """
+    if isinstance(github_auth, GithubToken):
+        return Github(auth=Token(github_auth))
+
+    app_auth = AppAuth(app_id=github_auth.client_id, private_key=github_auth.private_key)
+    app_installation_auth = AppInstallationAuth(
+        app_auth=app_auth, installation_id=github_auth.installation_id
+    )
+    return Github(auth=app_installation_auth)
+
+
+def _iter_delivery_attempts(
+    github_client: Github, webhook_address: WebhookAddress
+) -> Iterator[_WebhookDeliveryAttempts]:
+    """Iterate over webhook delivery attempts.
+
+    Args:
+        github_client: The GitHub client used to interact with the Github API.
+        webhook_address: The data to identify the webhook.
+    """
+    webhook_origin = (
+        github_client.get_repo(f"{webhook_address.github_org}/{webhook_address.github_repo}")
+        if webhook_address.github_repo
+        else github_client.get_organization(webhook_address.github_org)
+    )
+    deliveries = webhook_origin.get_hook_deliveries(webhook_address.id)
+    for delivery in deliveries:
+        # we check that the API is really returning the expected fields with non-null vals
+        # as pygithub is not doing this validation for us
+        required_fields = {"id", "status", "delivered_at", "event"}
+        none_fields = {
+            field for field in required_fields if getattr(delivery, field, None) is None
+        }
+        if none_fields:
+            raise AssertionError(
+                f"The webhook delivery {delivery.raw_data} is missing required fields:"
+                f" {none_fields}"
+            )
```

Comment:
Do you think we should stop all redelivery attempts if a field is missing for one?

Reply:
Or this should never actually happen?